README.md
# React Trip Animation

[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=mschezes%40gmail%2ecom&lc=ES&item_name=Miguel%20S%c3%a1nchez&item_number=github&no_note=0&currency_code=EUR&bn=PP%2dDonationsBF%3abtn_donate_SM%2egif%3aNonHostedGuest) [![npm](https://img.shields.io/npm/v/react-trip-animation.svg?)](http://badge.fury.io/js/react-trip-animation) [![npm](https://img.shields.io/npm/dt/react-trip-animation.svg)]() [![npm](https://img.shields.io/npm/l/react-trip-animation.svg)](https://opensource.org/licenses/MIT)

## Demo

Check the [demo](https://mschez.github.io/react-trip-animation/public).


## Installation

You can install `react-trip-animation` by executing this command:

```
npm install --save react-trip-animation
```

When the installation process ends, you are ready to import `react-trip-animation` to your project.


## How to Use

To use `react-trip-animation` you have to import the component in your project:

```javascript
import { Destination, Journey, Origin, Trip } from 'react-trip-animation';

/* Barcelona to Doha, Doha to Hanoi */
const TripAnimation = () => (
  <Trip>
    <Journey>
      <Origin position={[41.3887901, 2.1589899]} />
      <Destination position={[25.286106, 51.534817]} />
    </Journey>
    <Journey>
      <Origin position={[25.286106, 51.534817]} />
      <Destination position={[21.028511, 105.804817]} />
    </Journey>
  </Trip>
);
```

You can add more properties to `react-trip-animation`, see the list of available properties in [Properties](#properties) section.

## Trip Component Properties

### className (*String*)
> Default: *''*

The class name to wrap the *Trip* component.

### delayStart (*Number* milliseconds)
> Default: *0*

Defines time to wait before start the animation.

### fitWorld (*Boolean*)
> Default: *false*

The animation will start with the map adjusted to the whole view of the world.

### flyDuration (*number* milliseconds)
> Default: *2000*

Defines the time spent to move between the different scenes (initial animation and the time between the different journeys animation).

### height (*String*)
> Default: *'100%'*

Height size. It will override the size specified in the *className* prop.

### in (*Boolean*)
> Default: *true*

This prop stops the inmediately starting. The animation will not start until *in* is true.

### mode (*String*: **Trip.mode.ALL** | **Trip.mode.SEGMENT**)
> Default: *Trip.mode.ALL*

It defines the whole animation scene:
- *Trip.mode.ALL*: The map view will fit to the bounds of all journeys and never will move.
- *Trip.mode.SEGMENT*: The map view will fit to the bounds of the current segment.

### reset (*Boolean*)
> Default: *true*

Allows to reset the animation on finish.

### tile (*String*)
> Default: *'https://{s}.basemaps.cartocdn.com/dark_nolabels/{z}/{x}/{y}{r}.png'*

Defines the map skin. It is based on the standard map tiles system. You can find your favourite [here](http://leaflet-extras.github.io/leaflet-providers/preview).

### width (*String*)
> Default: *'100%'*

Width size. It will override the size specified in the *className* prop.

## Journey Component Properties

### cubicBezier (*String*)
> Default: *'0.8 0.12 0.2 0.9'*

Describes the ease animation between the origin and destination. You can calculate it [here](http://franzheidl.github.io/keysplines/).

### deep (*Number*)
> Default: *8*

Defines the curvature of the path between the origin and destination.

### delay (*Number* milliseconds)
> Default: *2000*

Allows to add a time before start the animation for the journey.

### duration (*Number* milliseconds)
> Default: *4000*

Defines the journey duration.

### ghostPathColor (*String*)
> Default: *'white'*

Defines the color of the ghost path.

### ghostPathDashArray (*String*)
> Default: *'4,10'*

Defines the style of the ghost path (more [info](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/stroke-dasharray)).

### ghostPathWeight (*Number*)
> Default: *2*

Defines the weight of the ghost path.

### keepPathOnFinish (*Boolean*)
> Default: *true*

Allows to hide/show de path after finish the animation in the journey.

### pathColor (*String*)
> Default: *'red'*

Defines the color of the path.

### pathIcon (*Image* url|base64)
> Default: *null*

Defines the icon that follows the path (plane icon for instance).

### pathIconSize (*Number*)
> Default: *40*

Defines the size of the icon.

### pathWeight (*Number*)
> Default: *3*

Defines the weight of the path.

### roundSide (*String*: **Journey.roundSide.LEFT_ROUND** | **Journey.roundSide.RIGHT_ROUND**)
> Default: *null*

Allows to change the way of the curvature.

### showGhostPath (*Boolean*)
> Default: *false*

Allows to show a ghost path between the origin and destination.

## Origin and Destination Component Properties

### markerAnchor (*Array*)
> Default: *[]*,

The anchor position to center the marker on the right point.

### markerIcon (*Node|Element*)
> Default: *null*

A React element to set as marker.

### position (*Array*)
> Required

The coordinates of the place to set as origin or destination. [*lat*, *lng*].


## Roadmap

- [ ] Anotations
- [ ] Train journeys
- [ ] Boat journeys

CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v0.3.0](https://github.com/mschez/react-trip-animation/compare/v0.2.0...v0.3.0)

- Bump node-fetch from 2.6.0 to 2.6.1 [`#7`](https://github.com/mschez/react-trip-animation/pull/7)
- Bump http-proxy from 1.18.0 to 1.18.1 [`#6`](https://github.com/mschez/react-trip-animation/pull/6)
- Bump elliptic from 6.5.2 to 6.5.3 [`#3`](https://github.com/mschez/react-trip-animation/pull/3)
- Bump lodash from 4.17.15 to 4.17.19 [`#2`](https://github.com/mschez/react-trip-animation/pull/2)
- Bump websocket-extensions from 0.1.3 to 0.1.4 [`#1`](https://github.com/mschez/react-trip-animation/pull/1)
- Update dependencies [`0fa01d5`](https://github.com/mschez/react-trip-animation/commit/0fa01d5215b4ca5b15bc7858e3cd046467c22807)
- Implement Server Side Rendering [`293029d`](https://github.com/mschez/react-trip-animation/commit/293029dfbaae30e589b2fac125408ac2adb2b8f6)
- Fix typo [`7181643`](https://github.com/mschez/react-trip-animation/commit/7181643c74eda81db1a65592d530ebe0159507b6)

#### [v0.2.0](https://github.com/mschez/react-trip-animation/compare/v0.1.1...v0.2.0)

> 16 May 2020

- Remove old dist file [`bda8c99`](https://github.com/mschez/react-trip-animation/commit/bda8c99aa161db9ded527edc26b741369195d49d)
- Fix undefined icon [`9a2e52b`](https://github.com/mschez/react-trip-animation/commit/9a2e52bf684ba71bdaaf560213322d1f9dd224df)

#### [v0.1.1](https://github.com/mschez/react-trip-animation/compare/v0.1.0...v0.1.1)

> 15 May 2020

- Update dependencies [`ab1ca97`](https://github.com/mschez/react-trip-animation/commit/ab1ca97ca75178576d16e656e7878b9b8af3a479)
- Improve prod script [`5c28a61`](https://github.com/mschez/react-trip-animation/commit/5c28a61bfa50b47750abc0bf32050595f7d12a4c)
- Remove console.log [`730a183`](https://github.com/mschez/react-trip-animation/commit/730a1833b1959a57e8e55c409093623ed150d89d)

#### v0.1.0

> 15 May 2020

- Init repo [`9453433`](https://github.com/mschez/react-trip-animation/commit/94534339fcc5cda992e1d2780950d9c428019c81)


public/App.js
import React, { useState } from 'react';

import {
  Destination,
  Journey,
  Origin,
  Trip,
} from '../src';
import iconDark from './iconDark.png';
import iconLight from './iconLight.png';
import iconAccent from './iconAccent.png';
import Marker from './Marker';

// eslint-disable-next-line max-len
const TILE_DARK = 'https://{s}.basemaps.cartocdn.com/dark_nolabels/{z}/{x}/{y}{r}.png';
// eslint-disable-next-line max-len
const TILE_LIGHT = 'https://{s}.basemaps.cartocdn.com/light_nolabels/{z}/{x}/{y}{r}.png';
// eslint-disable-next-line max-len
const TILE_WATERCOLOR = 'https://stamen-tiles-{s}.a.ssl.fastly.net/watercolor/{z}/{x}/{y}.png';
// eslint-disable-next-line max-len
const TILE_TONER = 'https://stamen-tiles-{s}.a.ssl.fastly.net/toner-background/{z}/{x}/{y}{r}.png';

const journeyPredefinedStyle = {
  [TILE_DARK]: {
    pathIcon: iconLight,
  },
  [TILE_LIGHT]: {
    pathIcon: iconDark,
  },
  [TILE_TONER]: {
    pathIcon: iconAccent,
  },
  [TILE_WATERCOLOR]: {
    pathIcon: iconDark,
  },
};

const defaultTripState = {
  delayStart: 2000,
  fitWorld: false,
  flyDuration: 2000,
  height: '100%',
  mode: Trip.mode.SEGMENT,
  tile: TILE_DARK,
  width: '100%',
};

const defaultJurneyState = {
  cubicBezier: '0.8 0.12 0.2 0.9',
  deep: 8,
  delay: 2000,
  duration: 6000,
  ghostPathColor: 'white',
  ghostPathDashArray: '4,10',
  ghostPathWeight: 2,
  keepPathOnFinish: true,
  pathColor: 'turquoise',
  pathIcon: iconLight,
  pathWeight: 3,
  reset: true,
  roundSide: null,
  showGhostPath: true,
};

const JourneyDemo = () => {
  const [selected, setSelected] = useState(0);
  const [trip, setTrip] = useState(defaultTripState);
  const [barcelonaDoha, setBarcelonaDoha] = useState(defaultJurneyState);
  const [dohaHanoi, setDohaHanoi] = useState(defaultJurneyState);

  const journeys = [barcelonaDoha, dohaHanoi];
  const setters = [setBarcelonaDoha, setDohaHanoi];

  const handleTrip = (event) => {
    event.persist();

    let { value } = event.target;

    if (event.target.type === 'number') {
      value = +event.target.value;
    }

    if (event.target.type === 'checkbox') {
      value = event.target.checked;
    }

    if (
      event.target.name === 'tile'
    ) {
      setBarcelonaDoha((prev) => ({
        ...prev,
        ...journeyPredefinedStyle[value],
      }));
      setDohaHanoi((prev) => ({
        ...prev,
        ...journeyPredefinedStyle[value],
      }));
    }

    setTrip((prev) => ({
      ...prev,
      [event.target.name]: value,
    }));
  };

  const handleChangeJurney = (event) => {
    event.persist();

    let { value } = event.target;

    if (event.target.type === 'number') {
      value = +event.target.value;
    }

    if (event.target.type === 'checkbox') {
      value = event.target.checked;
    }

    setters[selected]((prev) => ({
      ...prev,
      [event.target.name]: value,
    }));
  };

  return (
    <div className="container">
      <div className="map">
        <Trip
          delayStart={trip.delayStart}
          fitWorld={trip.fitWorld}
          height={trip.height}
          key={JSON.stringify({ bd: barcelonaDoha, dh: dohaHanoi, t: trip })}
          mode={trip.mode}
          tile={trip.tile}
          width={trip.width}
        >
          <Journey
            deep={barcelonaDoha.deep}
            delay={barcelonaDoha.delay}
            duration={barcelonaDoha.duration}
            ghostPathColor={barcelonaDoha.ghostPathColor}
            ghostPathDashArray={barcelonaDoha.ghostPathDashArray}
            ghostPathWeight={barcelonaDoha.ghostPathWeight}
            keepPathOnFinish={barcelonaDoha.keepPathOnFinish}
            pathColor={barcelonaDoha.pathColor}
            pathIcon={barcelonaDoha.pathIcon}
            pathIconSize={40}
            pathWeight={barcelonaDoha.pathWeight}
            roundSide={barcelonaDoha.roundSide}
            showGhostPath={barcelonaDoha.showGhostPath}
          >
            <Origin
              markerAnchor={[16, 24]}
              markerIcon={<Marker />}
              position={[41.3887901, 2.1589899]}
            />
            <Destination
              markerAnchor={[16, 24]}
              markerIcon={<Marker />}
              position={[25.286106, 51.534817]}
            />
          </Journey>
          <Journey
            deep={dohaHanoi.deep}
            delay={dohaHanoi.delay}
            duration={dohaHanoi.duration}
            ghostPathColor={dohaHanoi.ghostPathColor}
            ghostPathDashArray={dohaHanoi.ghostPathDashArray}
            ghostPathWeight={dohaHanoi.ghostPathWeight}
            keepPathOnFinish={dohaHanoi.keepPathOnFinish}
            pathColor={dohaHanoi.pathColor}
            pathIcon={dohaHanoi.pathIcon}
            pathIconSize={40}
            pathWeight={dohaHanoi.pathWeight}
            roundSide={dohaHanoi.roundSide}
            showGhostPath={dohaHanoi.showGhostPath}
          >
            <Origin
              markerAnchor={[16, 24]}
              markerIcon={<Marker />}
              position={[25.286106, 51.534817]}
            />
            <Destination
              markerAnchor={[16, 24]}
              markerIcon={<Marker />}
              position={[21.028511, 105.804817]}
            />
          </Journey>
        </Trip>
      </div>
      <div className="controls">
        <h1>
          React Trip Animation
        </h1>
        <div className="explain">
          Demo props for
          {' '}
          <a href="https://github.com/mschez/react-trip-animation/">
            React Trip Animation
          </a>
        </div>
        <div className="panel">
          <h3>
            Trip
          </h3>
          <section>
            <div className="title">
              Tile
            </div>
            <div className="explain">
              The
              {' '}
              <strong>tile</strong>
              {' '}
              prop defines the map skin. It is based on the standard map
              tiles system. You can find your favourite
              {' '}
              <a
                href="http://leaflet-extras.github.io/leaflet-providers/preview"
              >
                here
              </a>
              {' '}
              (even you can build your own).
            </div>
            <div className="option-wrapper">
              <select
                name="tile"
                onChange={handleTrip}
                value={trip.tile}
              >
                <option value={TILE_DARK}>Dark</option>
                <option value={TILE_LIGHT}>Light</option>
                <option value={TILE_WATERCOLOR}>Watercolor</option>
                <option value={TILE_TONER}>Toner</option>
              </select>
            </div>
          </section>
          <section>
            <div className="title">
              Mode
            </div>
            <div className="explain">
              <strong>mode</strong>
              {' '}
              prop describes the way the map play with the diferent journeys
              defined in the map.
            </div>
            <div className="option-wrapper">
              <select
                name="mode"
                onChange={handleTrip}
                value={trip.mode}
              >
                <option value={Trip.mode.ALL}>ALL</option>
                <option value={Trip.mode.SEGMENT}>SEGMENT</option>
              </select>
            </div>
          </section>
          <section>
            <div className="title">
              Delay Start
            </div>
            <div className="explain">
              <strong>delayStart</strong>
              {' '}
              prop allows you to add time before the animation start
            </div>
            <div className="option-wrapper">
              <input
                name="delayStart"
                onChange={handleTrip}
                type="text"
                value={trip.delayStart}
              />
            </div>
          </section>
          <section>
            <div className="title">
              Fly Duration
            </div>
            <div className="explain">
              <strong>flyDuration</strong>
              {' '}
              prop configures the time of the initial animation and the
              time between the different journeys animation (fly animation)
            </div>
            <div className="option-wrapper">
              <input
                name="flyDuration"
                onChange={handleTrip}
                type="text"
                value={trip.flyDuration}
              />
            </div>
          </section>
          <section>
            <div className="title">
              Fit World
            </div>
            <div className="explain">
              <strong>fitWorld</strong>
              {' '}
              prop allows to change the initial point to start the animation to
              fit the world
            </div>
            <div className="option-wrapper">
              <label htmlFor="fitWorld-option">
                <input
                  checked={trip.fitWorld}
                  id="fitWorld-option"
                  name="fitWorld"
                  onChange={handleTrip}
                  type="checkbox"
                />
                Fit World
              </label>
            </div>
          </section>
          <section>
            <div className="title">
              Reset
            </div>
            <div className="explain">
              <strong>reset</strong>
              {' '}
              prop allows to reset the animation on finish
            </div>
            <div className="option-wrapper">
              <label htmlFor="fitWorld-option">
                <input
                  checked={trip.reset}
                  id="reset-option"
                  name="reset"
                  onChange={handleTrip}
                  type="checkbox"
                />
                Reset
              </label>
            </div>
          </section>
        </div>
        <div className="panel">
          <h3>
            Journeys
          </h3>
          <section>
            <div className="title">
              Journey
            </div>
            <div className="explain">
              Select the journey to configure the props individually
            </div>
            <div className="option-wrapper">
              <select
                onChange={(event) => setSelected(event.target.value)}
                value={selected}
              >
                <option value={0}>Barcelona - Doha</option>
                <option value={1}>Doha - Hanoi</option>
              </select>
            </div>
          </section>
          <section>
            <div className="title">
              Cubic Bezier
            </div>
            <div className="explain">
              <strong>cubicBezier</strong>
              {' '}
              describes the ease animation between the origin and destination.
              You can calculate it
              {' '}
              <a
                href="http://franzheidl.github.io/keysplines/"
              >
                here
              </a>
            </div>
            <div className="option-wrapper">
              <input
                name="cubicBezier"
                onChange={handleChangeJurney}
                type="text"
                value={journeys[selected].cubicBezier}
              />
            </div>
          </section>
          <section>
            <div className="title">
              Deep
            </div>
            <div className="explain">
              <strong>deep</strong>
              {' '}
              prop defines the curvature of the path between the origin and
              destination.
            </div>
            <div className="option-wrapper">
              <input
                name="deep"
                onChange={handleChangeJurney}
                type="number"
                value={+journeys[selected].deep}
              />
            </div>
          </section>
          <section>
            <div className="title">
              Delay
            </div>
            <div className="explain">
              <strong>delay</strong>
              {' '}
              prop allows to add a time before start the animation for the
              journey
            </div>
            <div className="option-wrapper">
              <input
                name="delay"
                onChange={handleChangeJurney}
                type="number"
                value={+journeys[selected].delay}
              />
            </div>
          </section>
          <section>
            <div className="title">
              Duration
            </div>
            <div className="explain">
              <strong>duration</strong>
              {' '}
              prop defines the journey duration
            </div>
            <div className="option-wrapper">
              <input
                name="duration"
                onChange={handleChangeJurney}
                type="number"
                value={+journeys[selected].duration}
              />
            </div>
          </section>
          <section>
            <div className="title">
              Show Ghost Path
            </div>
            <div className="explain">
              <strong>showGhostPath</strong>
              {' '}
              prop allows to show a ghost path.
            </div>
            <div className="option-wrapper">
              <label htmlFor="showGhostPath-option">
                <input
                  checked={journeys[selected].showGhostPath}
                  id="showGhostPath-option"
                  name="showGhostPath"
                  onChange={handleChangeJurney}
                  type="checkbox"
                />
                Show Ghost Path
              </label>
            </div>
          </section>
          {journeys[selected].showGhostPath && (
            <section>
              <div className="title">
                Ghost Path Color
              </div>
              <div className="explain">
                <strong>ghostPathColor</strong>
                {' '}
                prop defines the color of the ghost path
              </div>
              <div className="option-wrapper">
                <input
                  name="ghostPathColor"
                  onChange={handleChangeJurney}
                  type="text"
                  value={journeys[selected].ghostPathColor}
                />
              </div>
            </section>
          )}
          {journeys[selected].showGhostPath && (
            <section>
              <div className="title">
                Ghost Path Dash Array
              </div>
              <div className="explain">
                <strong>ghostPathDashArray</strong>
                {' '}
                prop defines the style of the ghost path
              </div>
              <div className="option-wrapper">
                <input
                  name="ghostPathDashArray"
                  onChange={handleChangeJurney}
                  type="text"
                  value={journeys[selected].ghostPathDashArray}
                />
              </div>
            </section>
          )}
          {journeys[selected].showGhostPath && (
            <section>
              <div className="title">
                Ghost Path Weight
              </div>
              <div className="explain">
                <strong>ghostPathWeight</strong>
                {' '}
                prop defines the weight of the ghost path
              </div>
              <div className="option-wrapper">
                <input
                  name="ghostPathWeight"
                  onChange={handleChangeJurney}
                  type="number"
                  value={+journeys[selected].ghostPathWeight}
                />
              </div>
            </section>
          )}
          <section>
            <div className="title">
              Keep path on finish
            </div>
            <div className="explain">
              <strong>keepPathOnFinish</strong>
              {' '}
              prop allows to hide/show de path after finish
            </div>
            <div className="option-wrapper">
              <label htmlFor="keepPathOnFinish-option">
                <input
                  checked={journeys[selected].keepPathOnFinish}
                  id="keepPathOnFinish-option"
                  name="keepPathOnFinish"
                  onChange={handleChangeJurney}
                  type="checkbox"
                />
                keepPathOnFinish
              </label>
            </div>
          </section>
          <section>
            <div className="title">
              Path Color
            </div>
            <div className="explain">
              <strong>pathColor</strong>
              {' '}
              props defines the color of the path
            </div>
            <div className="option-wrapper">
              <input
                name="pathColor"
                onChange={handleChangeJurney}
                type="text"
                value={journeys[selected].pathColor}
              />
            </div>
          </section>
          <section>
            <div className="title">
              Path Weight
            </div>
            <div className="explain">
              <strong>pathWeight</strong>
              {' '}
              props defines the weight of the path
            </div>
            <div className="option-wrapper">
              <input
                name="pathWeight"
                onChange={handleChangeJurney}
                type="number"
                value={+journeys[selected].pathWeight}
              />
            </div>
          </section>
          <section>
            <div className="title">
              Round Side
            </div>
            <div className="explain">
              <strong>roundSide</strong>
              {' '}
              prop allows to change the way of the curvature
            </div>
            <div className="option-wrapper">
              <select
                name="roundSide"
                onChange={handleChangeJurney}
                value={trip.roundSide}
              >
                <option value={null}>none</option>
                <option value={Journey.roundSide.LEFT_ROUND}>Left</option>
                <option value={Journey.roundSide.RIGHT_ROUND}>Right</option>
              </select>
            </div>
          </section>
        </div>
      </div>
    </div>
  );
};

export default JourneyDemo;
